src/callbacks/summarize_images.cpp

```diff
@@ -0,0 +1,355 @@
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+//
+// summarize_images .hpp .cpp - Callback hooks to dump
+// results of image testing to event files
+////////////////////////////////////////////////////////////////////////////////
+
+#include <lbann_config.hpp>
+#include "lbann/callbacks/summarize_images.hpp"
+
+#include <lbann/layers/io/input/generic_input_layer.hpp>
+#include <lbann/proto/helpers.hpp>
+#include <lbann/utils/factory.hpp>
+#include <lbann/utils/image.hpp>
+#include <lbann/utils/summary.hpp>
+
+#include <callbacks.pb.h>
+
+#include <iostream>
+
+namespace lbann {
+namespace callback {
+
+// Strategy construction
+namespace {
+
+// Define the factory type.
+using factory_type = lbann::generic_factory<
+  image_output_strategy,
+  std::string,
+  proto::generate_builder_type<image_output_strategy,
+                               google::protobuf::Message const&>,
+  default_key_error_policy>;
+
+void register_default_builders(factory_type& factory) {
+  factory.register_builder("CategoricalAccuracyStrategy",
+                           build_categorical_accuracy_strategy_from_pbuf);
+  factory.register_builder("AutoencoderStrategy",
+                           build_autoencoder_strategy_from_pbuf);
+}
+
+// Manage a global factory
+struct factory_manager {
+  factory_type factory_;
+
+  factory_manager() {
+    register_default_builders(factory_);
+  }
+};
+
+factory_manager factory_mgr_;
+factory_type const& get_strategy_factory() noexcept {
+  return factory_mgr_.factory_;
+}
+
+std::unique_ptr<image_output_strategy>
+construct_strategy(google::protobuf::Message const& proto_msg) {
+  auto const& factory = get_strategy_factory();
+  auto const& msg =
+    proto::helpers::get_oneof_message(proto_msg, "strategy_type");
+  return factory.create_object(msg.GetDescriptor()->name(), msg);
+}
+
+}// namespace (Strategy construction)
+
+
+// categorical_accuracy_strategy
+std::vector<std::pair<size_t, El::Int>>
+categorical_accuracy_strategy::get_image_indices(model const& m) const {
+  static size_t img_counter = 0;
+  static size_t epoch_counter = 0;
+  auto const& exe_ctx = dynamic_cast<sgd_execution_context const&>(m.get_execution_context());
+  if(exe_ctx.get_epoch() > epoch_counter){
+    epoch_counter++;
+    img_counter = 0;
+  }
+  std::vector<std::pair<size_t, El::Int>> img_indices;
+
+  auto const& cat_accuracy_layer = get_layer_by_name(m, m_cat_accuracy_layer_name);
+
+  const BaseDistMat& categorized_correctly_dist =
+    cat_accuracy_layer.get_activations(*(cat_accuracy_layer.get_child_layers().front()));
+  auto const& distdata = categorized_correctly_dist.DistData();
+  CircMat<El::Device::CPU> categorized_correctly(
+    *(distdata.grid), distdata.root);
+  El::Copy(categorized_correctly_dist, categorized_correctly);
+
+  if (categorized_correctly.Height() != El::Int(1))
+    LBANN_ERROR("Tom was wrong about this matrix. Oops.");
```

Comment:
```suggestion
    LBANN_ERROR("categorical_accuracy_strategry expected to find a tensor of size 1, ",
                "but found a tensor of size ",categorized_correctly.Height());
```
I anticipate this error will be common if users pass the wrong layer into the `categorical_accuracy_strategy`. A more descriptive error message would be useful.